File: smart-civic-system/c_algorithms/LinkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// A node in the doubly linked list representing a Comment
typedef struct CommentNode {
    int comment_id;
    char text[256];
    char author[50];
    struct CommentNode* prev;
    struct CommentNode* next;
} CommentNode;

// Doubly Linked List Structure
typedef struct {
    CommentNode* head;
    CommentNode* tail;
    int size;
} LinkedList;

// Initialize an empty list
void initList(LinkedList* list) {
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

// Append a new comment to the end of the list
void appendComment(LinkedList* list, int id, const char* text, const char* author) {
    CommentNode* newNode = (CommentNode*)malloc(sizeof(CommentNode));
    newNode->comment_id = id;
    strcpy(newNode->text, text);
    strcpy(newNode->author, author);
    newNode->next = NULL;
    
    if (list->tail == NULL) {
        // First element
        newNode->prev = NULL;
        list->head = newNode;
        list->tail = newNode;
    } else {
        // Appending to the tail
        newNode->prev = list->tail;
        list->tail->next = newNode;
        list->tail = newNode;
    }
    list->size++;
}

// Print comments in chronological sequence
void printChronological(LinkedList* list) {
    printf("Comments (Oldest to Newest):\n");
    CommentNode* curr = list->head;
    while (curr != NULL) {
        printf("[%s]: %s\n", curr->author, curr->text);
        curr = curr->next;
    }
    printf("\n");
}

// Print comments in reverse chronological sequence
void printReverseChronological(LinkedList* list) {
    printf("Comments (Newest to Oldest):\n");
    CommentNode* curr = list->tail;
    while (curr != NULL) {
        printf("[%s]: %s\n", curr->author, curr->text);
        curr = curr->prev;
    }
    printf("\n");
}

// Mock Database Connection for Presentation Purposes
typedef struct {
    int id;
    char text[256];
    char author[50];
} CommentData;

void fetchCommentsFromDatabase(CommentData* db, int* count) {
    printf("[DB] Connecting to PostgreSQL database...\n");
    printf("[DB] Executing Query: SELECT comment_id, text, author FROM Comments WHERE issue_id = 12 ORDER BY created_at ASC;\n");
    
    db[0] = (CommentData){1, "There is a massive pothole here.", "Citizen_A"};
    db[1] = (CommentData){2, "I tripped over it yesterday!", "Citizen_B"};
    db[2] = (CommentData){3, "Municipal team has been dispatched.", "Admin_User"};
    *count = 3;
    
    printf("[DB] Fetched %d comments successfully.\n\n", *count);
}

int main() {
    LinkedList list;
    initList(&list);

    CommentData db[100];
    int commentCount = 0;
    
    // Simulate fetching dynamic data from backend
    fetchCommentsFromDatabase(db, &commentCount);

    for (int i = 0; i < commentCount; i++) {
        appendComment(&list, db[i].id, db[i].text, db[i].author);
    }

    printChronological(&list);
    printReverseChronological(&list);

    return 0;
}

File: smart-civic-system/backend/algorithms/LinkedList.js
/**
 * LinkedList - Doubly Linked List
 * Usage: Comment chain ordering — each new comment is appended to the list.
 * Supports O(1) append, O(n) traversal, and deletion by commentId.
 */
class ListNode {
  constructor(data) {
    this.data = data; // comment object { comment_id, user_id, comment, created_at, ... }
    this.prev = null;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /** Append a comment to the end of the list (chronological order) */
  append(data) {
    const node = new ListNode(data);
    if (!this.tail) {
      this.head = this.tail = node;
    } else {
      node.prev = this.tail;
      this.tail.next = node;
      this.tail = node;
    }
    this.length++;
    return node;
  }

  /** Prepend a comment to the front */
  prepend(data) {
    const node = new ListNode(data);
    if (!this.head) {
      this.head = this.tail = node;
    } else {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    }
    this.length++;
    return node;
  }

  /** Delete a node by comment_id */
  deleteById(commentId) {
    let current = this.head;
    while (current) {
      if (current.data.comment_id === commentId) {
        if (current.prev) current.prev.next = current.next;
        else this.head = current.next;
        if (current.next) current.next.prev = current.prev;
        else this.tail = current.prev;
        this.length--;
        return true;
      }
      current = current.next;
    }
    return false;
  }

  /** Return all comments as an array (oldest → newest) */
  toArray() {
    const result = [];
    let current = this.head;
    while (current) { result.push(current.data); current = current.next; }
    return result;
  }

  /** Return all comments newest → oldest */
  toArrayReverse() {
    const result = [];
    let current = this.tail;
    while (current) { result.push(current.data); current = current.prev; }
    return result;
  }

  /** Build a LinkedList from an array of comment objects */
  static fromArray(comments) {
    const list = new LinkedList();
    for (const c of comments) list.append(c);
    return list;
  }
}

module.exports = LinkedList;
